Source/Editor/DualityEditor/Forms/ObjectRefSelectionDialog.cs

```diff
@@ -0,0 +1,281 @@
+
+			public Component ComponentReference
+			{
+				get { return this._componentReference; }
+				set { this._componentReference = value; }
+			}
+		}
+
+		public Color PathColor { get; set; }
+		public Type FilteredType { get; set; }
+
+		public string ResourcePath { get; set; }
+
+		public IContentRef ResourceReference { get; set; }
+		public GameObject GameObjectReference { get; set; }
+		public Component ComponentReference { get; set; }
+
+		private TreeModel Model { get; set; }
+
+		public ObjectRefSelectionDialog()
+		{
+			InitializeComponent();
+
+			this.Model = new TreeModel();
+
+			this.PathColor = Color.Gray;
+
+			this.objectReferenceListing.SelectionChanged += this.ResourceListingOnSelectedIndexChanged;
+			this.objectReferenceListing.DoubleClick += this.ResourceListingOnDoubleClick;
+
+			this.txtFilterInput.KeyDown += this.TxtFilterInputOnKeyDown;
+
+			this.nodeName.DrawText += this.NodeName_OnDrawText;
+			this.nodePath.DrawText += this.NodePath_OnDrawText;
+		}
+
+		private void TxtFilterInputOnKeyDown(object sender, KeyEventArgs keyEventArgs)
+		{
+			TreeNodeAdv tmp = null;
+
+			if (keyEventArgs.KeyCode == Keys.Down)
+			{
+				tmp = this.objectReferenceListing.SelectedNode.NextNode;
+			}
+			else if (keyEventArgs.KeyCode == Keys.Up)
+			{
+				tmp = this.objectReferenceListing.SelectedNode.PreviousNode;
+			}
+
+			if (tmp != null)
+			{
+				this.objectReferenceListing.SelectedNode = tmp;
+			}
+		}
+
+		private void NodePath_OnDrawText(object sender, DrawTextEventArgs drawTextEventArgs)
+		{
+			ReferenceNode node = drawTextEventArgs.Node.Tag as ReferenceNode;
+
+			if (node == null)
+			{
+				return;
+			}
+
+			drawTextEventArgs.TextColor = this.PathColor;
+		}
+
+		private void NodeName_OnDrawText(object sender, DrawTextEventArgs drawTextEventArgs)
+		{
+			ReferenceNode node = drawTextEventArgs.Node.Tag as ReferenceNode;
+
+			if (node == null)
+			{
+				return;
+			}
+
+			drawTextEventArgs.TextColor = this.objectReferenceListing.ForeColor;
+		}
+
+		protected override void OnShown(EventArgs e)
+		{
+			base.OnShown(e);
+
+			this.objectReferenceListing.ClearSelection();
+			this.objectReferenceListing.Model = this.Model;
+
+			this.objectReferenceListing.BeginUpdate();
+			this.Model.Nodes.Clear();
+			
+			if (this.FilteredType.IsSubclassOf(typeof(GameObject)) || this.FilteredType == typeof(GameObject))
+			{
+				this.Text = "Select a GameObject";
+
+				foreach (GameObject currentObject in Scene.Current.AllObjects)
+				{
+					ReferenceNode tmpNode = new ReferenceNode(currentObject);
+
+					this.Model.Nodes.Add(tmpNode);
+				}
+			}
+			else if (this.FilteredType.IsSubclassOf(typeof(Component)) || this.FilteredType == typeof(Component))
+			{
+				this.Text = string.Format("Select a {0} Component", this.FilteredType.GetTypeCSCodeName());
+
+				foreach (Component currentComponent in Scene.Current.FindComponents(this.FilteredType))
+				{
+					ReferenceNode tmpNode = new ReferenceNode(currentComponent);

```

Comment:
Suggestion: This is also true for other places in code, but why is this variable prefixed with a `tmp`? It isn't really any more temporary than any other local variable, so might as well skip the prefix and give it a regular name. `node` would be just fine here as long as the context is clear and there's no ambiguity in scope.